ocfweb/docs/docs/staff/backend/kubernetes.md

```diff
@@ -0,0 +1,204 @@
+cluster with *M* > 1 masters adds one more node of fault tolerance.  Therefore,
+adding an extra node to an odd numbered cluster gives us nothing. If interested
+read more [here][failure-tolerance].
+
+### Workers
+
+Workers are the brawn in the Kubernetes cluster. While master nodes are
+constantly sharing data, managing the control plane (routing inside the
+Kubernetes cluster), and scheduling services, workers primarily run
+[pods][pod].  `kubelet` is the service that executes pods as dictated by the
+control plane, performs health checks, and recovers from pod failures should
+they occur.  Workers also run an instance of `kube-proxy`, which forwards
+control plane traffic to the correct `kubelet`.
+
+### Pods
+
+In the Kubernetes world, pods are the smallest computing unit. A pod is made up
+of one or more containers. The difference between a pod and a standalone
+container is best illustrated by an example. Consider [ocfweb][ocfweb]; it is
+composed of several containers—the web container, static container, and worker
+container.  In Kubernetes, together these three containers form one pod, and it
+is pods that can be scaled up or down. A failure in any of these containers
+indicates a failure in the entire pod. An astute reader might wonder: _if pods
+can be broken down into containers, how can pods possibly be the smallest
+unit?_ Do note that if one wished to deploy a singleton container, it would
+still need to be wrapped in the pod abstraction for Kubernetes to do anything
+with it.
+
+While pods are essential for understanding Kubernetes, when writing services we
+don't actually deal in pods but one further abstraction,
+[deployments][deployment], which create pods for us.
+
+## OCF Kubernetes Cluster Bootstrapping
+
+Since almost all OCF architecture is bootstapped using Puppet, it was necessary
+for us to do the same with Kubernetes. We rely on the
+[puppetlabs-kubernetes][kubernetes-module] module to handle initial
+bootstrapping and bolt OCF specific configurations on top of it.
+`puppetlabs-kubernetes` performs two crucial tasks:
+
+1. Installs `etcd`, `kubelet`, `kube-proxy`, and `kube-dns`, initializes the
+   cluster, and applies a networking backend.
+2. Generates the required [PKI for Kubernetes and etcd][kubernetes-pki].
+
+Do note that `puppetlabs-kubernetes` is still very much a work in progress. If
+you notice an issue in the module you are encouraged to write a patch and send
+it upstream.
+
+### Kubernetes PKI
+
+All the private keys and certs for the PKI are in the puppet private share, in
+`/opt/puppet/shares/private/kubernetes`. We won't go into detail of everything
+contained there, but Kubernetes and `etcd` communication is authenticated using
+client certificates. All the necessary items for workers are included in
+`os/Debian.yaml`, although adding a new master to the cluster requires
+re-running [kubetool][puppetlabs-kubetool] to generate new `etcd server` and
```

Comment:
How is kubetool re-run? Is this manually or automatically, where should it be run from and what arguments should it be given?

(not sure if all of this should be answered here, but I think a bit more detail on actually using kubetool would be good)

Reply:
The docs I linked explain how to run it (it's a docker container that can be run on any host) and what parameters to pass. I think `kubetool` warrants it's own set of documentation. One thing you pointed out that I missed is that it does need to be run manually.